tests/test_modules.cpp

```diff
@@ -0,0 +1,463 @@
+            packet_count++;
+            VLOG(10) << "Sinking " << input << std::endl;
+        });
+
+        builder.make_edge(source_mod->output_port("source"), sink);
+    };
+
+    m_pipeline->make_segment("SimpleModule_Segment", init_wrapper);
+
+    auto options = std::make_shared<Options>();
+    options->topology().user_cpuset("0-1");
+    options->topology().restrict_gpus(true);
+
+    Executor executor(options);
+    executor.register_pipeline(std::move(m_pipeline));
+    executor.start();
+    executor.join();
+
+    EXPECT_EQ(packet_count, 42);
+}
+
+TEST_F(SegmentTests, ModuleAsSinkTest)
+{
+    using namespace modules;
+
+    unsigned int packet_count{0};
+
+    auto init_wrapper = [&packet_count](segment::Builder& builder) {
+        auto source = builder.make_source<bool>("source", [&packet_count](rxcpp::subscriber<bool>& sub) {
+            if (sub.is_subscribed())
+            {
+                for (unsigned int i = 0; i < 43; ++i)
+                {
+                    sub.on_next(true);
+                    packet_count++;
+                }
+            }
+
+            sub.on_completed();
+        });
+
+        auto sink_mod = builder.make_module<SinkModule>("ModuleSinkTest_mod1");
+
+        builder.make_edge(source, sink_mod->input_port("sink"));
+    };
+
+    m_pipeline->make_segment("SimpleModule_Segment", init_wrapper);
+
+    auto options = std::make_shared<Options>();
+    options->topology().user_cpuset("0-1");
+    options->topology().restrict_gpus(true);
+
+    Executor executor(options);
+    executor.register_pipeline(std::move(m_pipeline));
+    executor.start();
+    executor.join();
+
+    EXPECT_EQ(packet_count, 43);
+}
+
+TEST_F(SegmentTests, ModuleChainingTest)
+{
+    using namespace modules;
+
+    auto sink_mod     = std::make_shared<SinkModule>("ModuleChainingTest_mod2");
+    auto init_wrapper = [&sink_mod](segment::Builder& builder) {
+        auto config = nlohmann::json();
+        unsigned int source_count{42};
+        config["source_count"] = source_count;
+
+        auto source_mod = builder.make_module<SourceModule>("ModuleChainingTest_mod1", config);
+        builder.init_module(sink_mod);
+
+        builder.make_dynamic_edge<bool, bool>(source_mod->output_port("source"), sink_mod->input_port("sink"));
+    };
+
+    m_pipeline->make_segment("SimpleModule_Segment", init_wrapper);
+
+    auto options = std::make_shared<Options>();
+    options->topology().user_cpuset("0-1");
+    options->topology().restrict_gpus(true);
+
+    Executor executor(options);
+    executor.register_pipeline(std::move(m_pipeline));
+    executor.start();
+    executor.join();
+
+    EXPECT_EQ(sink_mod->m_packet_count, 42);
+}
+
+TEST_F(SegmentTests, ModuleNestingTest)
+{
+    using namespace modules;
+
+    unsigned int packet_count{0};
+
+    auto init_wrapper = [&packet_count](segment::Builder& builder) {
+        auto nested_mod = builder.make_module<NestedModule>("ModuleNestingTest_mod1");
```

Comment:
I agree that its more succinct and easier to read, but my contention is that having more robust name-spacing makes things easier to troubleshoot, especially in dynamic environments like Python.

In practice, no one should ever see or be aware of the full namespaced name outside of an error dump, they'll only interact with the input/output ports of the module by their simple, relative name.

I agree that the default `[segment_module]` isn't terribly informative, so it may make sense to force a module creator to supply a unique module type name. But I think removing the additional namespacing information introduces more cognative load on the troubleshooting side.

Reply:
Based on our discussion, updated so that:

- There is no longer a default module type prefix
- Prefix is now of the form: type_name<ModuleType>()::<module_name>